File: src/main/scala/ClassProcessor.scala
package org.creativelabs.seams

import java.io._
import japa.parser.ast._
import japa.parser.ast.visitor._
import japa.parser.ast.body._
import japa.parser.ast.`type`._
import japa.parser._
import scala.collection.JavaConversions._
import scala.collection._

class ClassProcessor(typeDeclaration: ClassOrInterfaceDeclaration) {
    private var _fields: Map[String, String] = findFields(typeDeclaration)

    private var _dependencies: Map[String, Set[String]] = new mutable.HashMap
    private var _dependenciesUponType: Map[String, Set[String]] = new mutable.HashMap

    def fields(): Map[String, String] = _fields

    def dependencies(): Map[String, Set[String]] = this._dependencies

    def dependenciesUponType(): Map[String, Set[String]] = this._dependenciesUponType

    def compute() {            
        processMethods(typeDeclaration, _fields)
    }

    def outData() {
        ScalaApp.printFields(_fields)
        ScalaApp.printDeps("Dependencies", _dependencies)
        ScalaApp.printDeps("UponType", _dependenciesUponType)
    }

    private def processMethods(n: ClassOrInterfaceDeclaration, 
                    classFields: Map[String, String]) {

        for (bd <- n.getMembers()) bd match {
            case md: MethodDeclaration =>
                val (deps, uponType) = findOutgoingDependencies(md, classFields)
                _dependencies += (md.getName -> deps)
                _dependenciesUponType += (md.getName -> uponType)
            case _ => 
        }
        
    }

    private def findOutgoingDependencies(md: MethodDeclaration, 
                    classFields: Map[String, String]): (Set[String], Set[String]) = {
        val body = md.getBody
        val dependencyCounter = new DependencyCounterVisitor(classFields)
        dependencyCounter.visit(body, null)
        (dependencyCounter.getDependencies, dependencyCounter.getDependenciesUponType)
    }

    private def findFields(typeDeclaration: ClassOrInterfaceDeclaration): 
                    Map[String, String] = {

        val fields = new mutable.HashMap[String, String]
        for (bd <- typeDeclaration.getMembers) bd match {
            case fd: FieldDeclaration => 
                for (vardecl <- fd.getVariables()) {
                    val tv = new TypeVisitor
                    fd.getType().accept(tv, null)
                    fields.put(vardecl.getId().getName(), tv.name)
                }
            case _ =>
        }
        fields
    }

    private class TypeVisitor extends VoidVisitorAdapter[Object] {
        var name: String = _

        override def visit(n: ClassOrInterfaceType, a: Object) = name = n.getName()

        override def visit(n: PrimitiveType, a: Object) = name = n.getType().toString()
    }

}

File: src/main/java/org/creativelabs/ClassProcessor.java
package org.creativelabs;

import japa.parser.ast.body.*;
import japa.parser.ast.stmt.*;

import org.creativelabs.report.*;
import org.creativelabs.ssa.SimpleSsaForm;

import java.util.*;

class ClassProcessor {
    private ClassOrInterfaceDeclaration typeDeclaration;

    protected Map<String, Collection<Dependency>> dependencies = new HashMap<String, Collection<Dependency>>();

    private DependencyCounterVisitorBuilder dependencyCounterBuilder;

    protected HashMap<String, InternalInstancesGraph> internalInstances
            = new HashMap<String, InternalInstancesGraph>();
    protected Set<SimpleSsaForm> forms = new HashSet<SimpleSsaForm>();

    ClassProcessor(ClassOrInterfaceDeclaration typeDeclaration,
                   DependencyCounterVisitorBuilder dependencyCounterBuilder) {
        this.typeDeclaration = typeDeclaration;
        this.dependencyCounterBuilder = dependencyCounterBuilder;
    }

    public void buildReport(ReportBuilder reportBuilder) {
        reportBuilder.setDependencies(typeDeclaration.getName(), dependencies);
        reportBuilder.setInternalInstances(typeDeclaration.getName(), internalInstances);
    }

    public DependenciesChart getDependenciesChart() {
        DependenciesChart chart = new DependenciesChart();
        for (Map.Entry<String, InternalInstancesGraph> entry : internalInstances.entrySet()) {
            chart.addInternalInstancesCountForMethod(entry.getKey(), entry.getValue().toSet().size());
        }
        Collection<String> classDependencies = new HashSet<String>();
        for (Collection<Dependency> dependencySet : dependencies.values()) {
            for (Dependency dependency : dependencySet) {
                classDependencies.add(dependency.getType() + "");
            }
        }
        chart.addDependenciesCountForClass(typeDeclaration.getName(), classDependencies.size());
        return chart;
    }

    public Map<String, InternalInstancesGraph> getInternalInstances() {
        return internalInstances;
    }

    public Map<String, Collection<Dependency>> getDependencies() {
        return dependencies;
    }

    public Set<SimpleSsaForm> getForms() {
        return forms;
    }

    public void compute() {
        for (BodyDeclaration bd : typeDeclaration.getMembers()) {
            if (bd instanceof MethodDeclaration) {
                MethodDeclaration md = (MethodDeclaration) bd;
                findOutgoingDependencies(md);
                //findSsaForm(md);
            }
        }
    }

    void findOutgoingDependencies(MethodDeclaration md) {
        BlockStmt body = md.getBody();
        dependencyCounterBuilder.setConstructedMethodArguments(md);
        DependencyCounterVisitor dependencyCounterVisitor = dependencyCounterBuilder.build();
        dependencyCounterVisitor.visit(body, null);

        if (dependencyCounterVisitor.getDependencies() != null) {
            dependencies.put(md.getName(), dependencyCounterVisitor.getDependencies());
        }
        if (dependencyCounterVisitor.getInternalInstances() != null) {
            internalInstances.put(md.getName(), dependencyCounterVisitor.getInternalInstances());
        }
    }

    void findSsaForm(MethodDeclaration md) {
        SsaFormBuilderVisitor ssaFormBuilderVisitor = new SsaFormBuilderVisitor();
        SimpleSsaForm form = new SimpleSsaForm(md.getName(),
                ssaFormBuilderVisitor.visit(md, new VariablesHolder(new HashMap<String, Integer>())));
        forms.add(form);
    }

}

